### PR TITLE
spring security

### PR DESCRIPTION
**Pull request by BohdanZorii**

### CODE REVIEW

File: src/main/java/mate/zorii/bookstore/model/Role.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private RoleName role;
+
+    @Override
+    public String getAuthority() {
+        return "ROLE_" + role.name();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/Role.java:25]:
```suggestion
        return role.name();
```

File: src/main/java/mate/zorii/bookstore/model/User.java [hunk 2]
```diff
@@ -27,4 +38,21 @@ public class User {
     private String lastName;
     @Column(length = 500)
     private String shippingAddress;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/User.java:41]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/zorii/bookstore/model/User.java [hunk 3]
```diff
@@ -27,4 +38,21 @@ public class User {
...
+    @JoinTable(name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id"))
+    private Set<Role> roles = new HashSet<>();
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return roles.stream()
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/User.java:49]:
```suggestion
        return roles;
```

File: src/main/java/mate/zorii/bookstore/model/User.java [hunk 4]
```diff
@@ -27,4 +38,21 @@ public class User {
...
+                .map(role -> new SimpleGrantedAuthority(role.getAuthority()))
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public String getUsername() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/User.java:55]: Where other methods of UserDetails?

File: src/main/java/mate/zorii/bookstore/security/CustomUserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import mate.zorii.bookstore.repository.UserRepository;
+import org.hibernate.Hibernate;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/security/CustomUserDetailsService.java:14]:
```suggestion
```

File: src/main/java/mate/zorii/bookstore/security/CustomUserDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(username).orElseThrow(() ->
+                new UsernameNotFoundException("Can't find user by email " + username));
+        Hibernate.initialize(user.getRoles());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/security/CustomUserDetailsService.java:23]:
```suggestion
```

File: src/main/java/mate/zorii/bookstore/security/CustomUserDetailsService.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(username).orElseThrow(() ->
+                new UsernameNotFoundException("Can't find user by email " + username));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/security/CustomUserDetailsService.java:22]:
```suggestion
                new EntityNotFoundException("Can't find user by email " + username));
```

File: src/main/java/mate/zorii/bookstore/security/CustomUserDetailsService.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(username).orElseThrow(() ->
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/security/CustomUserDetailsService.java:21]:
```suggestion
        return userRepository.findByEmail(username).orElseThrow(() ->
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/mate/zorii/bookstore/model/User.java (hunk 4 above)
Comment (BohdanZorii) [reply src/main/java/mate/zorii/bookstore/model/User.java:55]: they're default

**Review by Elena-Bruyako — APPROVED** [review]
good jo

**Pull request merged**